File: framec_tests/golang/state_vars/state_vars._test.go
package state_vars

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestSingleVariable(t *testing.T) {
	sm := NewStateVars()
	x := sm.(*stateVarsStruct)
	assert.Equal(t, StateVarsState_A, x._compartment_.State)
	sm.X() // increment x
	sm.X() // increment x
	assert.Equal(t, 2, x._compartment_.StateVars["x"].(int))
}

func TestMultipleVariables(t *testing.T) {
	sm := NewStateVars()
	x := sm.(*stateVarsStruct)
	sm.Y()
	assert.Equal(t, StateVarsState_B, x._compartment_.State)
	assert.Equal(t, x._compartment_.StateVars["y"].(int), 10)
	assert.Equal(t, x._compartment_.StateVars["z"].(int), 100)
	sm.Y()
	sm.Y()
	sm.Z()
	sm.Y()
	assert.Equal(t, x._compartment_.StateVars["y"].(int), 13)
	assert.Equal(t, x._compartment_.StateVars["z"].(int), 101)

}

func TestVariablesAreReset(t *testing.T) {
	sm := NewStateVars()
	x := sm.(*stateVarsStruct)
	sm.X() // increment x
	sm.X() // increment x
	assert.Equal(t, x._compartment_.StateVars["x"].(int), 2)
	sm.Z() // transition to B
	sm.Z() // increment z
	sm.Y() // increment y
	sm.Z() // increment z
	assert.Equal(t, x._compartment_.StateVars["y"].(int), 11)
	assert.Equal(t, x._compartment_.StateVars["z"].(int), 102)
	sm.X() // transition to A
	assert.Equal(t, x._compartment_.StateVars["x"].(int), 0)
	sm.Y() // transition to B
	assert.Equal(t, x._compartment_.StateVars["y"].(int), 10)
	assert.Equal(t, x._compartment_.StateVars["z"].(int), 100)

}

File: framec_tests/cpp/state_vars/state_vars_test.cpp
#include <gtest/gtest.h>
#include "state_vars.cpp"

class StateVarsController : public StateVars {
public:
    StateVarsController() : StateVars() {}
};

class StateVarsControllerTest : public ::testing::Test {
protected:
    void SetUp() override {
        sm = new StateVarsController();
    }

    void TearDown() override {
        delete sm;
    }

    StateVarsController* sm;
};

TEST_F(StateVarsControllerTest, TestSingleVariable) {
    EXPECT_EQ("1", sm->state_info());
    sm->X(); // increment x
    sm->X(); // increment x
    EXPECT_EQ(2,  std::any_cast<int>(sm->_compartment_->stateVars["x"]));
}

TEST_F(StateVarsControllerTest, TestMultipleVariables) {
    sm->Y();
    EXPECT_EQ("2", sm->state_info());
    EXPECT_EQ(10, std::any_cast<int>(sm->_compartment_->stateVars["y"]));
    EXPECT_EQ(100, std::any_cast<int>(sm->_compartment_->stateVars["z"]));
    sm->Y();
    sm->Y();
    sm->Z();
    sm->Y();
    EXPECT_EQ(13, std::any_cast<int>(sm->_compartment_->stateVars["y"]));
    EXPECT_EQ(101, std::any_cast<int>(sm->_compartment_->stateVars["z"]));
}

TEST_F(StateVarsControllerTest, TestVariablesAreReset) {
    sm->X(); // increment x
    sm->X(); // increment x
    EXPECT_EQ(2, std::any_cast<int>(sm->_compartment_->stateVars["x"]));
    sm->Z(); // transition to B
    sm->Z(); // increment z
    sm->Y(); // increment y
    sm->Z(); // increment z
    EXPECT_EQ(11, std::any_cast<int>(sm->_compartment_->stateVars["y"]));
    EXPECT_EQ(102, std::any_cast<int>(sm->_compartment_->stateVars["z"]));
    sm->X(); // transition to A
    EXPECT_EQ(0, std::any_cast<int>(sm->_compartment_->stateVars["x"]));
    sm->Y(); // transition to B
    EXPECT_EQ(10, std::any_cast<int>(sm->_compartment_->stateVars["y"]));
    EXPECT_EQ(100, std::any_cast<int>(sm->_compartment_->stateVars["z"]));
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
